### PR TITLE
[mesheryctl] Rework mesheryctl system channel to confirm to latest design and make it context and compose aware. Fixes #2246

### LINKED ISSUE #2246: [mesheryctl] Display version and channel on system channel
#### Desired Behavior

`mesheryctl system channel` -> returns current channel and version with Usage
`mesheryctl system channel view` -> returns list of all context with their respective channels

#### Acceptance Tests
Test 1) **Show user details of current-context channel and usage. Do not confirm context.**
```
$ mesheryctl system channel
Channel: stable
Version: latest

Usage: ....
```

Test 2) **Show user details of current-context channel. Confirm context.**
```
$ mesheryctl system channel view
Context: local
Channel: stable
Version: latest
```

Test 3) **Show user details of all contexts’ channel**
```
$ mesheryctl system channel view -all
Context: local
Channel: stable
Version: latest

Context: test
Channel: edge
Version: latest
```

### PR DESCRIPTION
**Pull request by anirudhjain75**
**Description**

Modify channel command to use context instead of directly modifying the file, allow system start to also build changes from the version and channel of current context allowing reuse of code for switch in channel command. In struct change from yaml to mapstructure to conform to viper usage

This PR fixes #2246 #2128

### CODE REVIEW

**Review by leecalcote — APPROVED** [review]
Thanks, @anirudhjain75. #2246 calls for support of the `view` subcommand in it's acceptance tests.

Will you be adding support in this PR?

File: mesheryctl/internal/cli/root/system/channel.go [hunk 1]
```diff
@@ -15,28 +15,45 @@
 package system
```
Comment (leecalcote) [inline mesheryctl/internal/cli/root/system/channel.go:15]: Does `system update` need to be changed as well or is `update` completely reusing `start` functions?

File: mesheryctl/internal/cli/root/system/channel.go [hunk 2]
```diff
@@ -45,90 +62,55 @@ var channelCmd = &cobra.Command{
...
-				return
-			}
-
-			log.Info("Subscribed to edge channel.")
+			ContextContent.Channel = channelNameSeperated[0]
+			ContextContent.Version = "latest"
 
 		default:
-			currentChannel := getCurrentContextChannel()
+			currentChannel := mctlCfg.Contexts[focusedContext].Channel
```
Comment (leecalcote) [inline mesheryctl/internal/cli/root/system/channel.go:86]: When users try to switch to something other than stable or edge, they should be shown a message indicating that these are the only two channels available.

File: mesheryctl/internal/cli/root/system/channel.go [hunk 3]
```diff
@@ -45,90 +62,55 @@ var channelCmd = &cobra.Command{
...
 			}
-			log.Info("Successfully switched channel...")
 		}
 
+		return nil
 	},
 }
 
 func init() {
 	channelCmd.Flags().StringVarP(&channelSet, "set", "s", "", "Release channel to be set for Meshery and its adapters.")
```
Comment (leecalcote) [inline mesheryctl/internal/cli/root/system/channel.go:113]: The --version flag isn't supported here.

**Review by leecalcote — CHANGES_REQUESTED** [review]

File: mesheryctl/internal/cli/root/system/channel.go (hunk 3 above)
Comment (anirudhjain75) [reply mesheryctl/internal/cli/root/system/channel.go:113]: I will add it

Comment (anirudhjain75) [reply mesheryctl/internal/cli/root/system/channel.go:113]: Ohh Yea in this case I handled the case on if user specifies either stable or stable:version 
If you need a version flag seperately, I can get it done

**Comment by leecalcote** [comment]
@anirudhjain75 this is still WIP given our conversation on subcommands, right?

**Comment by anirudhjain75** [comment]
Give me 30 mins and extra time for dinner and it will not be in WIP.
I'm written the logic

**Comment by leecalcote** [comment]
Good deal. 👍

**Comment by anirudhjain75** [comment]
Examples to test out
`mesheryctl system channel` -> success
`mesheryctl system channel view` -> success
`mesheryctl system channel view -a` -> success
`mesheryctl system channel set` -> fail
`mesheryctl system channel set stable` -> success
`mesheryctl system channel set stable-version` -> success
`mesheryctl system channel switch` -> fail
`mesheryctl system channel switch stable` -> success
`mesheryctl system channel switch stable-version` -> success

Also you can try by passing name of context using -c flag to override current context temporarily

**Comment by leecalcote** [comment]
Unchecked test cases are **failing**.

- [ ] Acceptance Test 1 [failing]. No usage shown to users.
- [ ] Acceptance Test 2 [failing]. Current context not shown to users.
- [ ] Acceptance Test 3 [failing]. Format should be "Context: <context-name", not "<context-name> context".
- [x] Test: Invalid channel: [passing].
- [ ] Test: Invalid version format for stable: [failing]. `set` should only accept "latest" or the following regex: `v\d+\.\d+.\d+\-(?:alpha|beta|rc)-\d+`
- [ ] - [ ] Test: Invalid version format for stable: [failing]. `switch` should only accept "latest" or the following regex: `v\d+\.\d+.\d+\-(?:alpha|beta|rc)-\d+`
- [ ] Test: Invalid version format for edge: [failing]. `set` should only accept "latest".
- [ ] Test: Invalid version format for edge: [failing]. `switch` should only accept "latest".
- [ ] Test: `mesheryctl system start` with stable channel pinned to a version [failing]. The pinned version only applies to Meshery server, not to the adapters.
- [ ] Test: Docker not running: [failing]. `./mesheryctl system channel switch stable` has an unhandled error if Docker isn't running.
- [ ] Test: `-c` context override `./mesheryctl system channel switch stable -c <context>`: [failing]. `switch` command does not account for `-c`.
- [ ] Test: `./mesheryctl system channel view -c <context>`: [failing]. Output should include confirmation of "Context: <context>", since this is not the current context.
- [ ] Test: `././mesheryctl system channel view --all`: [failing]. _Current_ context should be indicated, so that the user understands which one of the listed contexts is the current context. Include extra line at end of output: `\n\nCurrent context: <context>`

@anirudhjain75, create unit tests for each of these cases and any others.

**Comment by leecalcote** [comment]
@anirudhjain75 please rename this PR to be descriptive of what is being delivered. Include [mesheryctl] at the start of the title.

De-duplicate your currently assigned issues.

**Review by navendu-pottekkat — COMMENTED** [review]
The channel_test.go LGTM. Just tried it locally. Will go through it again. I think it is possible to keep the test files in a separate folder. Though I am not really sure. If we can keep it as a separate folder, it would be great.

**Comment by anirudhjain75** [comment]
> 
> 
> The channel_test.go LGTM. Just tried it locally. Will go through it again. I think it is possible to keep the test files in a separate folder. Though I am not really sure. If we can keep it as a separate folder, it would be great.

Hi @navendu-pottekkat , The idea behind keeping both the main files and test files together is how its done in golang.

**Comment by navendu-pottekkat** [comment]
> Hi @navendu-pottekkat , The idea behind keeping both the main files and test files together is how its done in golang.

@anirudhjain75 Yes, I got that but I read somewhere that we can have the test files in a separate folder. But as you said the golang convention is to place it right by the actual file. So it works.

**Review by navendu-pottekkat — COMMENTED** [review]
@anirudhjain75 Nice work with the `channel_test`! I have tested it out and the code looks good too.

File: mesheryctl/internal/cli/root/system/channel_test.go [hunk 4]
```diff
@@ -0,0 +1,121 @@
...
+	logrus.SetOutput(b)
+	logrus.SetFormatter(&utils.OnlyStringFormatterForLogrus{})
+	SystemCmd.SetOut(b)
+}
+
+func BreakupFunc(t *testing.T) {
+	viewCmd.Flags().VisitAll(setFlagValueAsUndefined)
+}
+
+func setFlagValueAsUndefined(flag *pflag.Flag) {
```
Comment (leecalcote) [inline mesheryctl/internal/cli/root/system/channel_test.go:49]: Ideally, this is defined as a constant, not a function.

Comment (anirudhjain75) [reply mesheryctl/internal/cli/root/system/channel_test.go:49]: Visit all accepts a function with the specified format. That is why it's a function.

Comment (leecalcote) [reply mesheryctl/internal/cli/root/system/channel_test.go:49]: Clarification: the variable defined in ‘ setFlagValueAsUndefined’ doesn’t need to be set in a function, but can just be a declared constant

**Comment by leecalcote** [comment]
Test: User is missing config.yaml
Behavior: `channel view` output is blank even though next context was created.

```
$ mesheryctl system channel
Missing Meshery config file.
Create default config now? [y/n]? y
Default config file created at /Users/lee/.meshery/config.yaml
Context: 
Channel: 
Version: 

Usage:
  mesheryctl system channel [flags]
  mesheryctl system channel [command]

Available Commands:
  set         set release channel and version
  switch      switch release channel and version
  view        view release channel and version

Flags:
  -h, --help   help for channel

Global Flags:
      --config string    config file (default location is: /Users/lee/.meshery/config.yaml)
  -c, --context string   Override specified context with current context.
  -v, --verbose          verbose output

Use "mesheryctl system channel [command] --help" for more information about a command.
```

**Comment by leecalcote** [comment]
Test: User has a config.yaml with current context.
Behavior: `channel view` output is blank even though a context is available.

```
$ mesheryctl system channel
Missing Meshery config file.
Create default config now? [y/n]? y
Default config file created at /Users/lee/.meshery/config.yaml
Context: 
Channel: 
Version: 

Usage:
  mesheryctl system channel [flags]
  mesheryctl system channel [command]

Available Commands:
  set         set release channel and version
  switch      switch release channel and version
  view        view release channel and version

Flags:
  -h, --help   help for channel

Global Flags:
      --config string    config file (default location is: /Users/lee/.meshery/config.yaml)
  -c, --context string   Override specified context with current context.
  -v, --verbose          verbose output

Use "mesheryctl system channel [command] --help" for more information about a command.
```

**Comment by anirudhjain75** [comment]
The failing test case is because of Order mismatch when iterating order the array which is strange 🤔

**Comment by leecalcote** [comment]
> The failing test case is because of Order mismatch when iterating order the array which is strange 🤔. Other than that everything is fine

Interesting. Seems surmountable. How close are you to tying this one off?

**Comment by anirudhjain75** [comment]
I'm confused on how to get it done efficiently

**Comment by leecalcote** [comment]
@anirudhjain75 does this PR consider for the centralized reading from `config.yaml` and refreshing of structs in config.go?

**Comment by anirudhjain75** [comment]
Yes

**Comment by leecalcote** [comment]
For the time being, are you ok with disabling this particular unit test and opening an issue to follow up on?

**Comment by anirudhjain75** [comment]
Sure I think it's possible.

**Comment by leecalcote** [comment]
Good deal. Merging!

**Review by leecalcote — APPROVED** [review]

**Pull request merged**